### ISSUE TITLE
 Authentication fails with New-Device-Based adaptive auth script template

### ISSUE THREAD

**Issue by mefarazath** [wso2/product-is#11230]
**Describe the issue:**
When an app is configured with the New-Device-Based adaptive auth script template, authentication fail error response is sent after the second-factor authentication(TOTP). Tried in safari, chrome and firefox.

Below stacktrace can be see in the logs.
```
[2021-02-10 00:22:31,229] [7809056f-e0de-4019-9590-71ff888a38a8] ERROR {org.wso2.carbon.identity.application.authentication.framework.config.model.graph.JsGraphBuilder} - Error in executing the javascript for service provider : MyApp, Javascript Fragment :
function (context) {
                            setCookie(context.response, cookieName, subject.identifier, {
                                'sign': true,
                                'max-age': deviceRememberPeriod
                            });
                        } java.lang.NullPointerException
	at org.wso2.carbon.tomcat.ext.transport.CarbonResponseWrapper.generateCookieString(CarbonResponseWrapper.java:101)
	at org.wso2.carbon.tomcat.ext.transport.CarbonResponseWrapper.addCookie(CarbonResponseWrapper.java:355)
	at org.wso2.carbon.tomcat.ext.transport.ResponseFacadeWrapper.addCookie(ResponseFacadeWrapper.java:36)
	at javax.servlet.http.HttpServletResponseWrapper.addCookie(HttpServletResponseWrapper.java:60)
	at javax.servlet.http.HttpServletResponseWrapper.addCookie(HttpServletResponseWrapper.java:60)
	at org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsServletResponse.addCookie(JsServletResponse.java:89)
	at org.wso2.carbon.identity.conditional.auth.functions.http.CookieFunctionImpl.setCookie(CookieFunctionImpl.java:113)
	at jdk.nashorn.internal.scripts.Script$Recompilation$12$19A$\^eval\_.L:1(<eval>:2)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:639)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.ScriptObjectMirror.call(ScriptObjectMirror.java:117)
	at org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl.GraphBasedSequenceHandler.lambda$executeFunction$1(GraphBasedSequenceHandler.java:677)
	at org.wso2.carbon.identity.application.authentication.framework.config.model.graph.JsGraphBuilder$JsBasedEvaluator.evaluate(JsGraphBuilder.java:961)
	at org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl.GraphBasedSequenceHandler.executeFunction(GraphBasedSequenceHandler.java:677)
	at org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl.GraphBasedSequenceHandler.handleDecisionPoint(GraphBasedSequenceHandler.java:639)
	at org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl.GraphBasedSequenceHandler.handleNode(GraphBasedSequenceHandler.java:171)
	at org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl.GraphBasedSequenceHandler.handle(GraphBasedSequenceHandler.java:126)
	at org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.DefaultAuthenticationRequestHandler.handle(DefaultAuthenticationRequestHandler.java:183)
	at org.wso2.carbon.identity.application.authentication.framework.handler.request.impl.DefaultRequestCoordinator.handle(DefaultRequestCoordinator.java:263)
	at org.wso2.carbon.identity.application.authentication.framework.servlet.CommonAuthenticationServlet.doPost(CommonAuthenticationServlet.java:53)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.eclipse.equinox.http.helper.ContextPathServletAdaptor.service(ContextPathServletAdaptor.java:37)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.wso2.carbon.tomcat.ext.servlet.DelegationServlet.service(DelegationServlet.java:68)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.wso2.carbon.identity.captcha.filter.CaptchaFilter.doFilter(CaptchaFilter.java:66)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:88)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.wso2.carbon.tomcat.ext.filter.CharacterSetFilter.doFilter(CharacterSetFilter.java:65)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:108)
	at org.wso2.carbon.tomcat.ext.valves.SameSiteCookieValve.invoke(SameSiteCookieValve.java:38)
	at org.wso2.carbon.identity.cors.valve.CORSValve.invoke(CORSValve.java:89)
	at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:110)
	at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:115)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:99)
	at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62)
	at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:145)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
	at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:57)
	at org.wso2.carbon.tomcat.ext.valves.RequestEncodingValve.invoke(RequestEncodingValve.java:49)
	at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:126)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

If the samesite attribute is not sent it results in an NPE at [tomcat response wrapper](https://github.com/wso2/carbon-kernel/blob/d15ca12d142ce4219ff2380eaa7a9926a3ca142d/core/org.wso2.carbon.tomcat.ext/src/main/java/org/wso2/carbon/tomcat/ext/transport/CarbonResponseWrapper.java#L101).

To fix,
There are few places to fix
- Fix adaptive template to set the samesite attribute
- Fix the adaptive function that sets the cookie to set a default value for samesite attribute
- Fix kernel to avoid NPE and default to Strict if not same site attribute is set.

**How To Reproduce:**
Create an app
Add New-Device-Based adaptive auth script template from "Script based configuration"
Try login in to the app.
** Sample app can be used to try this out.

**Expected behavior:**
Authentication flow should work without an issue.

**Device Information** (_Please complete the following information_) **:**
 - Device: MacBook Pro
 - OS: MacOS 10.13.6
 - Browser: Safari(13.1.2),Chrome(88.0.4324.146),Firefox(85.0)

**Suggested labels:**

**Comment by somindatommy** [comment]
All the PRs are merged. Therefore, closing the issue.

**Comment by mifrazmurthaja** [comment]
With the fix: https://github.com/wso2/carbon-kernel/pull/2921, the default value of samesite attribute will be `STRICT`. Won't this impact the `deviceAuth` cookie? If it'll be impacted, we should also set the samesite attribute explicitly for the `deviceAuth` cookie in the template (https://github.com/wso2/carbon-identity-framework/blob/master/features/application-mgt/org.wso2.carbon.identity.application.mgt.server.feature/resources/identity/authenticationTemplates/unseenDevices.json)

**Comment by mefarazath** [comment]
> With the fix: [wso2/carbon-kernel#2921](https://github.com/wso2/carbon-kernel/pull/2921), the default value of samesite attribute will be `STRICT`. Won't this impact the `deviceAuth` cookie? If it'll be impacted, we should also set the samesite attribute explicitly for the `deviceAuth` cookie in the template (https://github.com/wso2/carbon-identity-framework/blob/master/features/application-mgt/org.wso2.carbon.identity.application.mgt.server.feature/resources/identity/authenticationTemplates/unseenDevices.json)

Doesn't the adaptive template allow setting the same site attribute?
The fallback logic in kernel was to default to STRICT if not same site attribute is set for the cookie.

**Comment by mifrazmurthaja** [comment]
> Doesn't the adaptive template allow setting the same site attribute?
> The fallback logic in kernel was to default to STRICT if not same site attribute is set for the cookie.

It allows, but since this is one of the default template available, we should set the applicable value for the samesite attribute in the template itself (instead of adding it manually). This is fixed in master: https://github.com/wso2/carbon-identity-framework/blob/1a57eb5ee6ce35e546201b17e06ac8571c40e48c/features/application-mgt/org.wso2.carbon.identity.application.mgt.server.feature/resources/identity/authenticationTemplates/unseenDevices.json#L85
